core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushMergeWriterUpdateIntoConnector.java

```diff
@@ -0,0 +1,153 @@
+
+    public PushMergeWriterUpdateIntoConnector(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer, Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+    }
+
+    private final Metadata metadata;
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    @Override
+    public Pattern<TableFinishNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TableFinishNode node, Captures captures, Context context)
+    {
+        MergeWriterNode mergeWriter = captures.get(MERGE_WRITER_NODE_CAPTURE);
+        MergeProcessorNode mergeProcessor = captures.get(MERGE_PROCESSOR_NODE_CAPTURE);
+        ProjectNode project = captures.get(PROJECT_NODE_CAPTURE);
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+
+        TableHandle tableHandle = mergeWriter.getTarget().getHandle();
+        Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(context.getSession(), tableHandle);
+
+        List<String> orderedColumnNames = mergeWriter.getTarget().getMergeParadigmAndTypes().getColumnNames();
+        List<? extends Node> mergeAssignments = project.getAssignments().get(mergeProcessor.getMergeRowSymbol()).getChildren();
+        ImmutableMap.Builder<ColumnHandle, Constant> assignmentsBuilder = ImmutableMap.builder();
+        for (int i = 0; i < orderedColumnNames.size(); i++) {
+            String columnName = orderedColumnNames.get(i);
+            Node assigmentNode = mergeAssignments.get(i);
+            if (isSymbol(assigmentNode)) {
+                // the column is not updated
+                continue;
+            }
+            if (!isConstant(assigmentNode)) {
+                return Result.empty();
+            }
+
+            ConnectorExpressionTranslator.ConnectorExpressionTranslation expressionTranslation = ConnectorExpressionTranslator.translateConjuncts(
+                    context.getSession(),
+                    ((Expression) assigmentNode),
+                    context.getSymbolAllocator().getTypes(),
+                    plannerContext,
+                    typeAnalyzer);
+
+            ConnectorExpression connectorExpression = expressionTranslation.connectorExpression();
+            // we don't support any expressions in update statements yet, only constants
+            // also in case of UPDATE ALL statement - row represented as base64 call
+            if (!(connectorExpression instanceof Constant)) {
+                return Result.empty();
+            }
+            assignmentsBuilder.put(columnHandles.get(columnName), (Constant) connectorExpression);
+        }
+        Map<ColumnHandle, Constant> assignments = assignmentsBuilder.buildOrThrow();
```

Comment:
you may consider moving a code block above that builds assignments to separate method that would be extracted below.

Reply:
refactored